src/radical/pilot/agent/executing/popen.py

```diff
@@ -462,10 +534,226 @@ def _check_running(self):
+        ret += 'export RP_GTOD="%s"\n'          % self.gtod
+        ret += 'export RP_PROF="%s"\n'          % self.prof
+
+        if self._prof.enabled:
+            ret += 'export RP_PROF_TGT="%s/%s.prof"\n' % (sbox, tid)
+        else:
+            ret += 'unset  RP_PROF_TGT\n'
+
+        if 'RP_APP_TUNNEL' in os.environ:
+            ret += 'export RP_APP_TUNNEL="%s"\n' \
+                    % os.environ['RP_APP_TUNNEL']
+
+        return ret
+
+
+    # --------------------------------------------------------------------------
+    # exec
+    #
+    def _get_task_env(self, task, launcher):
+
+        ret = ''
+        td  = task['description']
+
+        # named_env's are prepared by the launcher
+        if td['named_env']:
+            ret += '\n# named environment\n'
+            ret += '. %s\n' % launcher.get_task_env(td['named_env'])
+
+        # also add any env vars requested in the task description
+        if td['environment']:
+            ret += '\n# task env settings\n'
+            for key,val in td['environment'].items():
+                ret += 'export %s="%s"\n' % (key, val)
+
+        return ret
+
+
+    # --------------------------------------------------------------------------
+    #
+    def _get_pre_exec(self, task):
+
+        ret  = ''
+
+        cmds = task['description']['pre_exec']
+        for cmd in cmds:
+            ret += '%s || (echo "pre_exec failed"; false) || exit 1\n' \
+                   % cmd
+
+        return ret
+
+
+    # --------------------------------------------------------------------------
+    #
+    def _get_rank_ids(self, n_ranks, launcher):
+
+        ret  = ''
+        ret += 'export RP_RANKS=%s\n' % n_ranks
+        ret += launcher.get_rank_cmd()
+
+        return ret
+
+
+    # --------------------------------------------------------------------------
+    #
+    def _get_post_exec(self, task):
+
+        ret  = ''
+        cmds = task['description']['post_exec']
+        for cmd in cmds:
+            ret += '%s || (echo "post_exec failed"; false) || exit 1\n' \
+                   % cmd
+
+        return ret
+
+
+    # --------------------------------------------------------------------------
+    #
+    # rank
+    #
+    def _get_pre_rank(self, rank_id, cmds=[]):
+
+        ret = ''
+        for cmd in cmds:
+            # FIXME: exit on error, but don't stall other ranks on sync
+            ret += '        %s\n' % cmd
+
+        return ret
+
+
+    # --------------------------------------------------------------------------
+    #
+    def _get_rank_sync(self, sig):
+
+        # FIXME: make sure that all ranks are alive
+        ret  = 'while test $(cat %s.sig | wc -l) -lt $RP_RANKS; do\n' % sig
+        ret += '    sleep 1\n'
+        ret += 'done\n'
+
+        return ret
+
+
+    # --------------------------------------------------------------------------
+    #
+    def _get_rank_exec(self, task, rank_id, rank, launcher):
```

Comment:
consistency in method names: `get_launch_cmd` -> `get_exec_cmd`? (+ `get_rank_setup`?)

Reply:
would like to keep for now.